kernel/aster-nix/src/syscall/utimens.rs

```diff
@@ -2,90 +2,244 @@
+// Function to update the access and modification times of a file by its file descriptor
+fn do_utimes_fd(fd: FileDesc, times: Option<Times>, flags: UtimensFlags) -> Result<SyscallReturn> {
+    // Check if flags are empty, otherwise return error
+    if !flags.is_empty() {
+        return_errno_with_message!(Errno::EINVAL, "invalid flags")
     }
     let current = current!();
     let dentry = {
-        let path = path.to_string_lossy();
-        if path.is_empty() {
-            return_errno_with_message!(Errno::ENOENT, "path is empty");
+        let fs_path = FsPath::new(fd, "")?;
+        let fs = current.fs().read();
+        fs.lookup(&fs_path)?
+    };
+
+    vfs_utimes(&dentry, times)
+}
+
+// Function to update the access and modification times of a file by its path
+fn do_utimes_path(
+    dirfd: FileDesc,
+    pathname: &str,
+    times: Option<Times>,
+    flags: UtimensFlags,
+) -> Result<SyscallReturn> {
+    let current = current!();
+    let dentry = {
+        if pathname.is_empty() {
+            return_errno_with_message!(Errno::ENOENT, "pathname is empty");
         }
-        let fs_path = FsPath::new(dirfd, path.as_ref())?;
+        // Determine the file system path and the corresponding entry
+        let fs_path = FsPath::new(dirfd, pathname)?;
         let fs = current.fs().read();
         if flags.contains(UtimensFlags::AT_SYMLINK_NOFOLLOW) {
             fs.lookup_no_follow(&fs_path)?
         } else {
             fs.lookup(&fs_path)?
         }
     };
-    if let Some(time) = atime {
-        dentry.set_atime(Duration::from(time));
-    }
-    if let Some(time) = mtime {
-        dentry.set_mtime(Duration::from(time));
+
+    vfs_utimes(&dentry, times)
+}
+
+// Common function to handle updating file times, supporting both fd and path based operations
+fn do_utimes(
+    dirfd: FileDesc,
+    pathname_ptr: Vaddr,
+    times: Option<Times>,
+    flags: u32,
+) -> Result<SyscallReturn> {
+    let flags = UtimensFlags::from_bits(flags)
+        .ok_or(Error::with_message(Errno::EINVAL, "invalid flags"))?;
+
+    // Decide whether to treat the operation as fd-based or path-based
+    if pathname_ptr == 0 || dirfd != AT_FDCWD {
+        do_utimes_fd(dirfd, times, flags)
+    } else {
+        let pathname = read_cstring_from_user(pathname_ptr, MAX_FILENAME_LEN)?;
+        let pathname = pathname.to_string_lossy();
+        do_utimes_path(dirfd, pathname.as_ref(), times, flags)
     }
-    Ok(SyscallReturn::Return(0))
+}
+
+// Sets the access and modification times for a file,
+// specified by a pathname relative to the directory file descriptor `dirfd`.
+fn do_futimesat(
+    dirfd: FileDesc,
+    pathname_ptr: Vaddr,
+    timespecs_ptr: Vaddr,
+) -> Result<SyscallReturn> {
+    let times = if timespecs_ptr != 0 {
+        let (mut autime, mut mutime) = read_timespec_from_user(timespecs_ptr)?;
+        if autime.nsec >= 1000000 || autime.nsec < 0 || mutime.nsec >= 1000000 || mutime.nsec < 0 {
```

Comment:
I think the current implementation is wrong. There are 2 `futimens` syscall, `sys_futimens` uses `timespec` while `sys_futimesat` uses `timeval`. So the  `sys_futimesat` should handle the type transform.

Reply:
On Linux, `__kernel_old_timeval` for `sys_futimesat` is composed of `__kernel_long_t` and `__kernel_suseconds_t`, which are corresponded to `long` and `unsigned long`.  Therefore, I think handling the value transform of seconds is enough.